include/libtorrent/announce_entry.hpp

```diff
@@ -107,20 +97,73 @@ namespace libtorrent {
 
 		int scrape_downloaded = -1;
 
+		// the number of times in a row we have failed to announce to this
+		// tracker.
+		std::uint8_t fails : 7;
+
+		// true while we're waiting for a response from the tracker.
+		bool updating : 1;
+
+		// set to true when we get a valid response from an announce
+		// with event=started. If it is set, we won't send start in the subsequent
+		// announces.
+		bool start_sent : 1;
+
+		// set to true when we send a event=completed.
+		bool complete_sent : 1;
+
+		// internal
+		bool triggered_manually : 1;
+
+		// reset announce counters and clears the started sent flag.
+		// The announce_endpoint will look like we've never talked to
+		// the tracker.
+		void reset();
+
+		// updates the failure counter and time-outs for re-trying.
+		// This is called when the tracker announce fails.
+		void failed(int backoff_ratio, seconds32 retry_interval = seconds32(0));
+
+		// returns true if we can announce to this tracker now.
+		// The current time is passed in as ``now``. The ``is_seed``
+		// argument is necessary because once we become a seed, we
+		// need to announce right away, even if the re-announce timer
+		// hasn't expired yet.
+		bool can_announce(time_point now, bool is_seed, std::uint8_t fail_limit) const;
+
+		// returns true if the last time we tried to announce to this
+		// tracker succeeded, or if we haven't tried yet.
+		bool is_working() const { return fails == 0; }
+	};
+
+	// this class holds information about one bittorrent tracker, as it
+	// relates to a specific torrent.
+	struct TORRENT_EXPORT announce_entry
+	{
+		// constructs a tracker announce entry with ``u`` as the URL.
+		explicit announce_entry(string_view u);
+		announce_entry();
+		~announce_entry();
+		announce_entry(announce_entry const&);
+		announce_entry& operator=(announce_entry const&);
+
+		// tracker URL as it appeared in the torrent file
+		std::string url;
+
+		// the current ``&trackerid=`` argument passed to the tracker.
+		// this is optional and is normally empty (in which case no
+		// trackerid is sent).
+		std::string trackerid;
+
+		std::vector<announce_endpoint> endpoints;
```

Comment:
Moving `announce_entry` is not trivial because it gets passed through `torrent_handle::async_call` which will copy any value parameters. We would need to use generalized lambda capture to move the parameters. That is a C++14 feature, although I think it's supported by all compilers we care about.